### PR TITLE
#108 - recipe fields for sharing

### PR DESCRIPTION
**Pull request by patrick-hebner**
- Add fields for collaborators, readers and owner
- Add authorization methods to restrict access to recipes
- Restrict access to recipes
- Fix some box
- Add tests

### CODE REVIEW

File: server/src/app/utils/crypto.utils.ts [hunk 1]
```diff
@@ -16,8 +16,8 @@ export class CryptoUtils {
     }
   }
 
-  public comparePasswords = async (firstPw: string, secondPw: string): Promise<void> => {
-    const result = await bcrypt.compare(firstPw, secondPw);
+  public comparePasswords = async (firstPw: string, secondPw: string, pwIsToken: boolean): Promise<void> => {
+    const result = pwIsToken ? firstPw === secondPw : await bcrypt.compare(firstPw, secondPw);
```
Comment (flashback2k14) [inline server/src/app/utils/crypto.utils.ts]: why did you add this? i don't get your intention.

File: server/src/auth/handler/auth.handler.ts [hunk 2]
```diff
@@ -12,7 +12,8 @@ export class AuthHandler {
   public login = async (req: SocoboRequest, res: Response, next: NextFunction): Promise<void> => {
     try {
       const result = await this._getExtractedRequestBodyResult(req, "login(..)");
-      const loginResult = await this._authService.login(result);
+      const pwIsToken = req.query.token;
```
Comment (flashback2k14) [inline server/src/auth/handler/auth.handler.ts]: is `token` query param a boolean? what's the differenz if this flag is set?

File: server/test/helper/TestHelper.ts [hunk 3]
```diff
@@ -39,4 +39,29 @@ export class TestHelper {
       return error;
     }
   }
+
+  public static getTokenForEmailAndPassword = async (
```
Comment (flashback2k14) [inline server/test/helper/TestHelper.ts]: is it possible to add this to the existing getToken method?

File: server/test/50-recipe.route.test.ts [hunk 4]
```diff
@@ -2,20 +2,29 @@ process.env["NODE_ENV"] = "test";
 
 import { expect } from "chai";
 import * as mocha from "mocha";
+
```
Comment (flashback2k14) [inline server/test/50-recipe.route.test.ts]: could you please reorder / group your tests. it costs many time if you override the token so much. :-)

**Review by flashback2k14 — COMMENTED** [review]
LGTM except to login method extension

File: server/src/app/utils/crypto.utils.ts (hunk 1 above)
Comment (hetec) [reply server/src/app/utils/crypto.utils.ts]: I needed to add this since I don't know the password of the mock users. So I couldn't login with these useless to test the authorization of the recipe endpoints. If you tell me all password I can remove this again. This method allowed me to login directly with the token instead of the password.

File: server/src/auth/handler/auth.handler.ts (hunk 2 above)
Comment (hetec) [reply server/src/auth/handler/auth.handler.ts]: If the token query param is true I can login directly with a token instead of a password. See above comment

File: server/test/50-recipe.route.test.ts (hunk 4 above)
Comment (hetec) [reply server/test/50-recipe.route.test.ts]: No, unfortunately not. The main problem is the the test don't reset the state of the db before each test. So each test depends on the test before it since the state of the db was changed. For example I had problems to use a specific user id in the recipes because it was the id you deleted your last user test (The cause of this error was a really hard to find). May I can check if I can reorder some of them but I also need a lot of different users to test the authorization stuff.

But in general it would be a good practice to reset the db after each test. What do you think?

File: server/test/helper/TestHelper.ts (hunk 3 above)
Comment (hetec) [reply server/test/helper/TestHelper.ts]: Maybe. I didn't want to have a method with many parameter where some of them will never be used. If both is implemented in one function we always need to set some parameters to null?! I'm not sure what's the best way here. Should I try to combine these functions to one?

File: server/src/app/utils/crypto.utils.ts (hunk 1 above)
Comment (flashback2k14) [reply server/src/app/utils/crypto.utils.ts]: the user passwords are inside the `30-socobouser.route.test.ts`file

please remove this. i think this is a big security issue if we could login with the password hash value.

File: server/src/auth/handler/auth.handler.ts (hunk 2 above)
Comment (flashback2k14) [reply server/src/auth/handler/auth.handler.ts]: this is not a token. its the password hash value you use. or get i wrong?

File: server/test/50-recipe.route.test.ts (hunk 4 above)
Comment (flashback2k14) [reply server/test/50-recipe.route.test.ts]: yes, i have the same big issues inside the fooditem tests. how could we realize the db restore after each test? i think call `npm test` with `require('child_process').exec()` is not the best solution!? maybe the restoring of the db after each test cost many time. what if we use the before or after hook?

File: server/test/helper/TestHelper.ts (hunk 3 above)
Comment (flashback2k14) [reply server/test/helper/TestHelper.ts]: no, its okay. now i know why you use the token query param. but please remove this query param.

File: server/src/recipe/middleware/recipe.authorization.service.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import { NextFunction, Request, Response } from "express";
+import {
+  ApiError, ERRORS, SocoboRequest
+} from "../../app/index";
+
+import { DbExtension } from "../../db/interface/db-extension";
+import { SocoboUser } from "../../socobouser/index";
+import { Recipe } from "../index";
+
+export class RecipeAuthorizationService {
```
Comment (flashback2k14) [inline server/src/recipe/middleware/recipe.authorization.service.ts]: what think you about to make this as a middleware? it's not really needed, but it would looks better and you doesn't need the SocoboUserRequest inside your RecipeHandler!? :-)

File: server/test/50-recipe.route.test.ts (hunk 4 above)
Comment (flashback2k14) [reply server/test/50-recipe.route.test.ts]: we could implement the db setup programmatically inside the test folder similar i did [here](https://github.com/flashback2k14/playthatsong/blob/master/server/logic/services/setup.service.js)!?

Comment (hetec) [reply server/test/50-recipe.route.test.ts]: Yes, we should do it this way and then use beforeEach and afterEach to do the setup and the cleanup?!

Comment (hetec) [reply server/test/50-recipe.route.test.ts]: And I think because these tests are integration tests the speed is not the most important aspect because they are already a lot slower then unit test (because they use http an db connections and so on)

Comment (hetec) [reply server/test/50-recipe.route.test.ts]: We should prefer stable and correct test over the execution time in this case I think.

File: server/src/auth/handler/auth.handler.ts (hunk 2 above)
Comment (hetec) [reply server/src/auth/handler/auth.handler.ts]: I just call these things token since a token is simply a hashed value. I didn't mean the auth token here.

File: server/src/app/utils/crypto.utils.ts (hunk 1 above)
Comment (hetec) [reply server/src/app/utils/crypto.utils.ts]: Yes I will remove this, it was only to test my routes. I didn't plan to keep this in the code.

File: server/test/helper/TestHelper.ts (hunk 3 above)
Comment (hetec) [reply server/test/helper/TestHelper.ts]: ok

File: server/src/recipe/middleware/recipe.authorization.service.ts (hunk 5 above)
Comment (hetec) [reply server/src/recipe/middleware/recipe.authorization.service.ts]: I will check if its possible in all cases

Comment (hetec) [reply server/src/recipe/middleware/recipe.authorization.service.ts]: I'm not sure how I should do this without repeating a lot of code from the handler methods. Because I need to fetch the recipe before I can check if the user is authorized. So if I want to use a middleware I would need to fetch the recipe in the middleware and then check the authorization. After this I would need to fetch the recipe again in the handler to do the remaining work. Maybe I could add the recipe / recipes to the request, but there are case where I need different recipe data in the handler (e.g. getAllRecipes and search). Any Idea?

File: server/test/50-recipe.route.test.ts (hunk 4 above)
Comment (flashback2k14) [reply server/test/50-recipe.route.test.ts]: Yes, you are right. So think the beforeEach hook should create a fresh DB. AfterEach hook it's not needed than.

File: server/src/recipe/middleware/recipe.authorization.service.ts (hunk 5 above)
Comment (flashback2k14) [reply server/src/recipe/middleware/recipe.authorization.service.ts]: Yes you could add this to the request. But your solution is good. We should use it and move on.

**Comment by hetec** [comment]
So I hope it's done now

**Comment by flashback2k14** [comment]
You didn't committed anything. :-)

**Comment by hetec** [comment]
VS Code just didn't show me the error in the editor.

**Review by flashback2k14 — APPROVED** [review]
